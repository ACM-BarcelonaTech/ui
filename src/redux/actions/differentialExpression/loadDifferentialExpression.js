import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from 'redux/actionTypes/differentialExpression';

import { fetchWork } from 'utils/work/fetchWork';
import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';

const getCellSetName = (name) => (name?.split('/')[1] || name);

const loadDifferentialExpression = (
  experimentId, cellSets, comparisonType, tableState, newAdvancedFilters = null,
) => async (dispatch, getState) => {
  const advancedFilters = newAdvancedFilters
  ?? getState().differentialExpression.comparison.advancedFilters;

  dispatch({
    type: DIFF_EXPR_LOADING,
    payload: {
      experimentId,
      advancedFilters,
    },
  });

  const body = {
    name: 'DifferentialExpression',
    experimentId,
    cellSet: getCellSetName(cellSets.cellSet),
    compareWith: getCellSetName(cellSets.compareWith),
    basis: getCellSetName(cellSets.basis),
  };

  let pagination = {};
  if (tableState) {
    const currentPageSize = tableState.pagination.pageSize;
    pagination = {
      orderBy: tableState.sorter.field,
      orderDirection: (tableState.sorter.order === 'ascend') ? 'ASC' : 'DESC',
      offset: ((tableState.pagination.current - 1) * currentPageSize),
      limit: currentPageSize,
      responseKey: 0,
    };
    if (advancedFilters.length > 0) {
      pagination.filters = advancedFilters;
    }
    if (tableState.geneNamesFilter) {
      pagination.filters = [{
        columnName: 'gene_names',
        type: 'text',
        expression: tableState.geneNamesFilter,
      }];
    }
    pagination = { pagination };
  }

  const timeout = getTimeoutForWorkerTask(getState(), 'DifferentialExpression');
  try {
    const data = await fetchWork(
      experimentId, body, getState, { timeout, extras: pagination },
    );
    let { total } = data;
    const { rows } = data;
    if (!total && !Object.keys(pagination).length) {
      total = rows.length;
    }

    dispatch({
      type: DIFF_EXPR_LOADED,
      payload: {
        experimentId,
        data: rows,
        cellSets,
        total,
        comparisonType,
      },
    });
  } catch (error) {
    dispatch({
      type: DIFF_EXPR_ERROR,
      payload: {
        experimentId,
        error,
      },
    });
  }
};

export default loadDifferentialExpression;
