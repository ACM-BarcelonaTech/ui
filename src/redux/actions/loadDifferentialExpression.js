import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from '../actionTypes/differentialExpression';

import sendWork from '../../utils/sendWork';

const REQUEST_TIMEOUT = 60;

const loadDifferentialExpression = (
  experimentId, cellSets, tableState,
) => async (dispatch) => {
  dispatch({
    type: DIFF_EXPR_LOADING,
    payload: {
      experimentId,
    },
  });

  const currentPageSize = tableState.pagination.pageSize;

  const body = {
    name: 'DifferentialExpression',
    ...cellSets,
  };

  const pagination = {
    pagination: {
      orderBy: tableState.sorter.field,
      orderDirection: (tableState.sorter.order === 'ascend') ? 'ASC' : 'DESC',
      offset: ((tableState.pagination.current - 1) * currentPageSize),
      limit: currentPageSize,
      responseKey: 0,
    },
  };

  try {
    const response = await sendWork(experimentId, REQUEST_TIMEOUT, body, pagination);
    const data = JSON.parse(response.results[0].body);
    const { rows, total } = data;

    return dispatch({
      type: DIFF_EXPR_LOADED,
      payload: {
        experimentId,
        data: rows,
        cellSets,
        total,
      },
    });
  } catch (error) {
    console.error(error);
    dispatch({
      type: DIFF_EXPR_ERROR,
      payload: {
        experimentId,
        error: "Couldn't fetch differential expression results data.",
      },
    });
  }
};

export default loadDifferentialExpression;
