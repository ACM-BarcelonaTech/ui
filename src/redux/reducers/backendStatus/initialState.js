const initialExperimentBackendStatus = {
  loading: false,
  error: false,
  status: {
  },
};

const initialState = {};

export default initialState;

export { initialExperimentBackendStatus };
