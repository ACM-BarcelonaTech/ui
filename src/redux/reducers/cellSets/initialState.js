const initialState = {
  properties: {},
  hierarchy: [],
  selected: [],
  loading: true,
  error: false,
};

export default initialState;
