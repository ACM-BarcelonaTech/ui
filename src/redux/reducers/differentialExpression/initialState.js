const initialState = {
  properties: {
    data: [],
    comparisonGroup: {},
    comparisonType: null,
    loading: false,
    error: false,
    total: 0,
  },
  comparison: {
    type: 'within',
    advancedFilters: [],
    group: {
      between: {
        cellSet: null,
        compareWith: null,
        basis: null,
      },
      within: {
        cellSet: null,
        compareWith: null,
        basis: null,
      },
    },
    ordering: {
      orderBy: 'logFC',
      orderDirection: 'DESC',
    },
  },
};

export default initialState;
