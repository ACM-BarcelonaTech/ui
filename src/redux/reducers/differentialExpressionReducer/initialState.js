const initialState = {
  properties: {
    data: {},
    cellSets: {},
    loading: false,
    error: false,
  },
};

export default initialState;
