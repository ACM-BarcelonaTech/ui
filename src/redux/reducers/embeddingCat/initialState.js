const initialState = {
  width: 700,
  height: 550,
  pointSize: 5,
  toggleInvert: '#FFFFFF',
  masterColour: '#000000',
  umap1Domain: null,
  umap2Domain: null,
  titleText: '',
  titleSize: 20,
  titleAnchor: 'start',
  axisTitlesize: 13,
  axisTicks: 13,
  transGrid: 0,
  axesOffset: 10,
  masterFont: 'sans-serif',
  xaxisText: 'UMAP 1',
  yaxisText: 'UMAP 2',
  pointStyle: 'circle',
  pointOpa: 5,
  g1Color: 'red',
  g2mColor: 'green',
  sColor: 'blue',
  legendTextColor: '#FFFFFF',
  legendEnabled: true,
  legend: null,
  geneexpLegendloc: '',
  labelSize: 28,
  labelShow: 1,
  labelFont: 2,
  labelsEnabled: true,
  selectedClusters: [],
  testVar: null,
};

export default initialState;
