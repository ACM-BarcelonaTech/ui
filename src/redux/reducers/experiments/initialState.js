const experimentTemplate = {
  projectUuid: null,
  name: null,
  description: 'Analysis description',
  id: null,
  createdDate: null,
  lastViewed: null,
  notifyByEmail: false,
  meta: { organism: null, type: '10x' },
  sampleIds: [],
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
  },
  ids: [],
};

export default initialState;
export { experimentTemplate };
