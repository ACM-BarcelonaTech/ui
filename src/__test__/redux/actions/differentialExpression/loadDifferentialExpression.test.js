import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import loadDifferentialExpression from '../../../../redux/actions/differentialExpression/loadDifferentialExpression';
import initialState from '../../../../redux/reducers/differentialExpressionReducer/initialState';
import sendWork from '../../../../utils/sendWork';

import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from '../../../../redux/actionTypes/differentialExpression';

jest.mock('localforage');
jest.mock('../../../../utils/sendWork', () => ({
  __esModule: true, // this property makes it work
  default: jest.fn(),
}));

const mockStore = configureStore([thunk]);

describe('loadDifferentialExpression action', () => {
  const experimentId = '1234';
  const cellSets = {
    first: 'louvain-0',
    second: 'louvain-1',
  };

  it('Dispatches appropriately on failure', async () => {
    const store = mockStore({
      differentialExpression: {
        ...initialState,
      },
    });
    sendWork.mockImplementation(() => new Promise((resolve, reject) => reject(new Error('random error!'))));

    await store.dispatch(loadDifferentialExpression(experimentId, 'Versus rest', cellSets));
    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(DIFF_EXPR_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(DIFF_EXPR_ERROR);
    expect(loadedAction).toMatchSnapshot();
  });

  it('Dispatches appropriately on success', async () => {
    const store = mockStore({
      differentialExpression: {
        ...initialState,
      },
    });

    sendWork.mockImplementation(() => {
      const resolveWith = {
        results:
          [
            {
              body: JSON.stringify({
                rows: [
                  {
                    pval: 1.16, qval: 2.16, log2fc: 3.45, gene_names: 'A',
                  },
                  {
                    pval: 2.16, qval: 3.16, log2fc: 4.45, gene_names: 'B',
                  },
                  {
                    pval: 3.16, qval: 4.16, log2fc: 5.45, gene_names: 'C',
                  },
                  {
                    pval: 4.16, qval: 5.16, log2fc: 6.45, gene_names: 'D',
                  },
                  {
                    pval: 5.16, qval: 6.16, log2fc: 7.45, gene_names: 'E',
                  },
                  {
                    pval: 6.16, qval: 7.16, log2fc: 8.45, gene_names: 'F',
                  },
                ],
              }),
            },
          ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    await store.dispatch(loadDifferentialExpression(experimentId, 'Versus rest', cellSets));

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(DIFF_EXPR_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(DIFF_EXPR_LOADED);
    expect(loadedAction).toMatchSnapshot();
  });
});
