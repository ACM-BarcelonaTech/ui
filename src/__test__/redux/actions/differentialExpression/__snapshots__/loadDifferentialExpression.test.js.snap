// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadDifferentialExpression action Dispatches appropriately on failure 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on failure 2`] = `
Object {
  "payload": Object {
    "error": [Error: random error!],
    "experimentId": "1234",
  },
  "type": "diffExpr/error",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on success 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on success 2`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "condition/condition-control",
      "cellSet": "louvain/louvain-0",
      "compareWith": "louvain/louvain-1",
    },
    "comparisonType": "within",
    "data": Array [
      Object {
        "gene_names": "A",
        "logFC": 3.45,
        "p_val": 1.16,
        "p_val_adj": 2.16,
      },
      Object {
        "gene_names": "B",
        "logFC": 4.45,
        "p_val": 2.16,
        "p_val_adj": 3.16,
      },
      Object {
        "gene_names": "C",
        "logFC": 5.45,
        "p_val": 3.16,
        "p_val_adj": 4.16,
      },
      Object {
        "gene_names": "D",
        "logFC": 6.45,
        "p_val": 4.16,
        "p_val_adj": 5.16,
      },
      Object {
        "gene_names": "E",
        "logFC": 7.45,
        "p_val": 5.16,
        "p_val_adj": 6.16,
      },
      Object {
        "gene_names": "F",
        "logFC": 8.45,
        "p_val": 6.16,
        "p_val_adj": 7.16,
      },
    ],
    "experimentId": "1234",
    "total": undefined,
  },
  "type": "diffExpr/loaded",
}
`;
